sw/device/boot_rom2/ext_format.c

```diff
@@ -0,0 +1,130 @@
+ */
+typedef struct common_parse_result {
+  /**
+   * Absolute address of the start of the metadata region.
+   */
+  const uint8_t *metadata;
+  /**
+   * Length of the metadata region.
+   */
+  size_t metadata_len;
+  /**
+   * Absolute address of the start of the payload region.
+   */
+  const uint8_t *payload;
+  /**
+   * Length of the payload region.
+   */
+  size_t payload_len;
+  /**
+   * Absolute address of the next stage entrypoint.
+   */
+  void *entrypoint;
+} common_parse_result_t;
+
+/**
+ * Performs the "common" portion of parsing a ROM_EXT executable, mostly
+ * consiting of signature verification.
+ *
+ * @param binary a pointer to the binary.
+ * @param crypto a cryptographic vtable.
+ * @param version the version this parse is for (which is used to determine the
+ * metadata length).
+ * @return "common" parsed out values.
+ */
+static common_parse_result_t common_parse(const uint8_t *binary,
+                                          ext_crypto_vt_t crypto,
+                                          ext_format_version_t version) {
+  fixed_header_t header;
+  memcpy(&header, binary, sizeof(header));
+
+  // Compute offset pointers into |binary| pointing to relevant fields within.
+  const uint8_t *signature = binary + sizeof(header);
+  const uint8_t *footer_hash = signature + crypto.sig_len;
+  const uint8_t *metadata = footer_hash + crypto.hash_len;
+  size_t metadata_len = metadata_len_from_version(version);
+  const uint8_t *payload = metadata + metadata_len;
+
+  // Verify the magic word.
+  CHECK(header.magic == kExtMagic, "Magic value mismatch.");
+
+  // Hash the executable header, for signature verification.
+  const uint8_t *header_hash;
+  CHECK(crypto.hash_start(crypto.self), "Failed to start header hash.");
+  CHECK(crypto.hash_data(crypto.self, (const uint8_t *)&header, sizeof(header)),
+        "Failed to hash the fixed header.");
+  CHECK(crypto.hash_data(crypto.self, footer_hash, crypto.hash_len),
+        "Failed to hash the footer hash.");
+  CHECK(crypto.hash_finish(crypto.self, &header_hash),
+        "Failed to complete header hash.");
+
+  // Verify the header signature.
+  CHECK(crypto.verify_signature(crypto.self, header_hash, signature),
+        "Failed to verify header signature.");
+
+  // Hash the footer (i.e., everything after the footer hash).
+  const uint8_t *footer_hash_computed;
+  CHECK(crypto.hash_start(crypto.self), "Failed to start footer hash.");
+  CHECK(crypto.hash_data(crypto.self, metadata, metadata_len),
+        "Failed to hash metadata.");
+  CHECK(crypto.hash_data(crypto.self, payload, header.payload_len),
+        "Failed to hash payload.");
+  CHECK(crypto.hash_finish(crypto.self, &footer_hash_computed),
+        "Failed to complete footer hash.");
+
+  // NOTE: This comparison should probably be made constant-time, since we might
+  // not want an attacker to learn how close a failed hash is.
+  CHECKZ(memcmp(footer_hash, footer_hash_computed, crypto.hash_len),
```

Comment:
What is the benefit of adding an additional comparison instead of just applying the verify op on the overall hash?  This component is monolithic  anyway.

Reply:
I'm not sure I follow. Here's the scheme I'm using right now:
- The fixed-location components are authenticated by a signature, which can be verified without looking at any other values in the binary (besides reading them to put them into the HMAC engine, of course).
- The variable-length parts (the metadata + the payload) have a hash that can be computed using only values directly protected by the first signature. The hash itself is protected by the signature.

There isn't an overall signature. I've pushed a change that adds a high-level description of the format and its security model, which I think might give more context than a bunch of C code.

Reply:
OK, just for practical purposes assume that the flash page size is 4KB.  The ROM_EXT will be aligned to page size granularity to simplify handling of flash erase during firmware update etc. Thus in practice, the software image hash usually includes even the padding at the end of the image to the page size boundary.

The signature is also aligned to a 32bit multiple.  

Using this constraints, you can reduce the number of memory comparisons and simplify hardening. Usually what is done with monolithic components like this one is to aggregate all the hashes (SW, HW measurements, Other Measurements) using a cryptographic function, and  feed them into the verify function. 

The aggregated hash can then be used in a warm boot implementation that relies on a lighter verification scheme.